File: source/Evaluator.py
import mysql.connector
from datetime import datetime

class Evaluator():

    THRESHOLD_ADD = 0.6
    THRESHOLD_REVIEW = 0.3

    def __init__(self, cursor, another_cursor):
        self.cursor = cursor
        self.another_cursor = another_cursor

    def execute(self):
        categories = []

        try:
            self.cursor.execute("SELECT DISTINCT category FROM cs_join")
        except mysql.connector.Error as err:
            print("Query failed: {}".format(err))


        for row in self.cursor:
            category = row["category"]
            if category not in categories:
                categories.append(category)

        results = open('results' + str(datetime.now()).replace(' ', '').replace(':', '').replace('.', ''), 'w', encoding="utf8")
        results.write("Number of categories: " + str(len(categories)) + "\n")

        results.write("\nSubjects per category:\n")

        current_category = 1

        for category in categories:
            current_category += 1
            interim_result = self.get_probabilities(category, results)
            self.insert_suggestions(interim_result, category)


    def get_probabilities(self, category, results):
        probabilities = []

        # get number of distinct subjects in category
        try:
            query = "SELECT COUNT(DISTINCT subject) AS count FROM cs_join WHERE category = (%s)"
            self.cursor.execute(query, (category,))
        except mysql.connector.Error as err:
            print("Query failed: {}".format(err))
        subjects = 0
        for row in self.cursor:
            subjects = row["count"]

        results.write('#' + "\t" + category + "\t" + str(subjects) + "\n")

        # get distinct pairs of predicates and objects
        try:
            query = "SELECT DISTINCT predicate, object FROM cs_join WHERE category = (%s)"
            self.cursor.execute(query, (category,))
        except mysql.connector.Error as err:
            print("Query failed: {}".format(err))

        # count subjects that have the predicate object pair
        for row in self.cursor:
            predicate = row["predicate"]
            object = row["object"]
            try:
                query = "SELECT COUNT(*) AS count FROM cs_join WHERE category = (%s) AND predicate = (%s) AND object = (%s)"
                self.another_cursor.execute(query, (category, predicate, object))
            except mysql.connector.Error as err:
                print("Query failed: {}".format(err))
            concerned_subjects = 0
            for row in self.another_cursor:
                concerned_subjects = row["count"]

            # compute probability
            probability = int(concerned_subjects) / int(subjects)
            probabilities.append({"predicate": predicate, "object": object, "probability": probability})

        return probabilities

    def insert_suggestions(self, probabilities, category):
        for interim_result in probabilities:
            predicate = interim_result["predicate"]
            object = interim_result["object"]
            probability = interim_result["probability"]
            try:
                query = "SELECT subject FROM cs_join WHERE category = %s AND NOT predicate = %s AND NOT object = %s"
                self.cursor.execute(query, (category, predicate, object))
            except mysql.connector.Error as err:
                print("Query failed: {}".format(err))
            for row in self.cursor:
                subject = row["subject"]
                status = 'D'
                if probability > self.THRESHOLD_ADD:
                    status = 'A'
                elif probability > self.THRESHOLD_REVIEW:
                    status = 'R'
                try:
                    query = "INSERT INTO suggestions (status, subject, predicate, object, probability) VALUES (%s, %s, %s, %s, %s)"
                    self.another_cursor.execute(query, (status, subject, predicate, object, probability))
                except mysql.connector.Error as err:
                    print("Query failed: {}".format(err))

File: v0.2_php/source/Evaluator.php
<?php

include_once "./Helper.php";

class Evaluator {

	CONST THRESHOLD_ADD = 0.6;
	CONST THRESHOLD_REVIEW = 0.3;

	private $helper;

	public function __construct() {
		$this->helper = new Helper();
	}

	public function execute() {
		$db = $this->helper->getDB();

		echo "Preparing ...\n";
		$totalCategories = $db->query( 'SELECT COUNT(DISTINCT category) AS count FROM join_categories_statements' )->fetch_object()->count;

		$categories = array();
		$categoriesQuery = $db->query( 'SELECT DISTINCT category FROM join_categories_statements' );
		while ( $row = mysqli_fetch_row( $categoriesQuery ) ) {
			$categories[] = $row[0];
		}

		echo "Processing results ...\n";
		$evaluatedCategories = 1;

		$done = False;

		foreach ( $categories as $category ) {

			$interimResults = $this->getInterimResults( $db, $category );

			$this->createSuggestions( $db, $interimResults, $category );

			if ( !$done ) {
				$done = $this->helper->showProgress( $evaluatedCategories, $totalCategories );
			}
			
			$evaluatedCategories++;
		}
	}

	private function getInterimResults( $db, $category){
		$interimResult = array();
		$resourcesInCategoryQuery = $db->query( "SELECT COUNT(DISTINCT subject) AS count FROM join_categories_statements WHERE category = '$category'" );
		$resourcesInCategory = $resourcesInCategoryQuery->fetch_object()->count;
		$predicateObjectPairsQuery = $db->query( "SELECT DISTINCT predicate, object FROM join_categories_statements WHERE category = '$category'" );
		while ( $row = mysqli_fetch_row( $predicateObjectPairsQuery ) ) {
			$predicate = $row[0];
			$object = $row[1];
			$countQuery = $db->query( "SELECT COUNT(*) AS count FROM join_categories_statements WHERE category = '$category' AND predicate = '$predicate' AND object = '$object'" );
			$count = $countQuery->fetch_object()->count;
			$probability = (int)$count / (int)$resourcesInCategory;
			$interimResult[] = array(
				'predicate' => $predicate,
				'object' => $object,
				'probability' => $probability
			);
		}
		return $interimResult;
	}

	private function createSuggestions( $db, $interimResults, $category ) {
		foreach ( $interimResults as $result ) {
			$predicate = $result['predicate'];
			$object = $result['object'];
			$probability = $result['probability'];
			$itemsWithoutPairQuery = $db->query( "SELECT item FROM join_categories_statements WHERE category = '$category' AND NOT predicate = '$predicate' AND NOT object = '$object'" );
			if ( $itemsWithoutPairQuery ) {
				while ( $row = mysqli_fetch_row( $itemsWithoutPairQuery ) ) {
					$subject = $row[0];
					$status = 'D';
					if ( $probability > self::THRESHOLD_ADD ) {
						$status = 'A';

					} elseif ( $probability > self::THRESHOLD_REVIEW ) {
						$status = 'R';
					}
					$insert = "INSERT INTO suggestions (status, subject, predicate, object, probability) VALUES ('$status' '$subject', '$predicate', '$object', $probability)";
					$db->query( $insert );
				}
			}
		}
	}
}
